lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,182 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package log
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/gravitational/trace"
+)
+
+var (
+	// fileSharedWriter global file shared writer with the ability to reopen the log file
+	// in the same mode and with the same flags, either by function call or by filesystem
+	// notification subscription on the changes.
+	fileSharedWriter *FileSharedWriter
+
+	// initLock protects global init of the file shared writer.
+	initLock sync.Mutex
+)
+
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	watcher  atomic.Pointer[fsnotify.Watcher]
+	fileFlag int
+	fileMode fs.FileMode
+
+	lock sync.Mutex
+}
+
+// InitFileSharedWriter wraps the provided [os.File] in a writer that is thread safe,
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func InitFileSharedWriter(logFile *os.File, flag int, mode fs.FileMode) (io.Writer, error) {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		return nil, trace.BadParameter("file shared writer already initialized")
+	}
```

Comment:
Should a new invocation of `InitFileSharedWriter` just override the previous one instead of erroring out? Or can we rely on `applyLogConfig` only ever getting called once? I think that even during an in-process restart the configuration won't get reparsed, so we might be good here, but it's just a bit weird especially for tests.

Reply:
We always clear the state in applyConfig by `logutils.CloseFileSharedWriter()` then do `logutils.InitFileSharedWriter` only if in configuration we still have output as a file path. I've tested it with the SIGHUP signals, so we re-read the configuration and trigger applyConfig, also I've checked exitContext closing to clean up the state